spec/UseCaseValidationSpec.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Marcosh\LamPHPda\ValidationSpec;
+
+use Marcosh\LamPHPda\Instances\ListL\ConcatenationMonoid;
+use Marcosh\LamPHPda\Validation\Validation;
+use Marcosh\LamPHPda\Validation\Validation as V;
+
+final class TinValidator {
+    public static function isFollowBelgiumRule1(string $tin): bool
+    {
+        $divisionRemainderBy97 = (int) (substr($tin, 0, 9)) % 97;
+
+        return 97 - $divisionRemainderBy97 === (int) (substr($tin, 9, 3));
+    }
+
+    public static function isFollowBelgiumRule2(string $tin): bool
+    {
+        $divisionRemainderBy97 = (int) ('2' . substr($tin, 0, 9)) % 97;
+
+        return 97 - $divisionRemainderBy97 === (int) (substr($tin, 9, 3));
+    }
+
+    public static function getDateType(string $tin): int
+    {
+        $year = (int) (substr($tin, 0, 2));
+        $month = (int) (substr($tin, 2, 2));
+        $day = (int) (substr($tin, 4, 2));
+
+        $y1 = checkdate($month, $day, 1900 + $year);
+        $y2 = checkdate($month, $day, 2000 + $year);
+
+        if (0 === $day || 0 === $month || ($y1 && $y2)) {
+            return 3;
+        }
+
+        if ($y1) {
+            return 1;
+        }
+
+        if ($y2) {
+            return 2;
+        }
+
+        return 0;
+    }
+
+    public static function hasValidDateType(string $tin): bool
+    {
+        return 0 !== self::getDateType($tin);
+    }
+
+    public static function followRule1(string $tin): bool
+    {
+        $dateType = self::getDateType($tin);
+
+        return self::isFollowBelgiumRule1($tin) && (1 === $dateType || 3 === $dateType);
+    }
+
+    public static function followRule2(string $tin): bool
+    {
+        $dateType = self::getDateType($tin);
+
+        return self::isFollowBelgiumRule2($tin) && 2 <= $dateType;
+    }
+
+    public static function hasValidLength(string $tin): bool
+    {
+        return 11 === strlen($tin);
+    }
+
+    public static function hasValidPattern(string $tin): bool
+    {
+        $pattern = '\\d{2}[0-1]\\d[0-3]\\d{6}';
+
+        return 1 === preg_match(sprintf('/%s/', $pattern), $tin);
+    }
+
+    public static function Belgian(): Validation
+    {
+        $r0 = V::isString(['TIN type must be a string.'])
+            ->then(V::satisfies([TinValidator::class, 'hasValidLength'], ['TIN length is invalid.']))
+            ->then(V::satisfies([TinValidator::class, 'hasValidPattern'], ['TIN pattern is invalid.']))
+            ->then(V::satisfies([TinValidator::class, 'hasValidDateType'], ['TIN date is invalid.']));
+
+        $r1 = $r0->then(V::satisfies([TinValidator::class, 'followRule1'], ['TIN validation of rule 1 failed.']));
+        $r2 = $r0->then(V::satisfies([TinValidator::class, 'followRule2'], ['TIN validation of rule 2 failed.']));
+
+        return $r1->or(new ConcatenationMonoid(), $r2);
```

Comment:
wouldn't it work anyway doing

```
$r1 = V::satisfies([TinValidator::class, 'followRule1'], ['TIN validation of rule 1 failed.']);
$r2 = V::satisfies([TinValidator::class, 'followRule2'], ['TIN validation of rule 2 failed.']);

return $r0->then($r1->or(new ConcatenationMonoid(), $r2));
```

This would avoid the recomputation of `$r0` for the two branches

Reply:
Fixed.